server/src/main/java/org/elasticsearch/script/field/ToScriptField.java

```diff
@@ -0,0 +1,205 @@
+
+        }
+
+        public DocValuesField<?> getScriptField(SortedNumericDocValues sortedNumericDocValues, String name) {
+            return new DelegateDocValuesField(new Longs(sortedNumericDocValues), name);
+        }
+    }
+
+    public static class ToHalfFloatScriptField extends ToScriptField {
+
+        public static final ToHalfFloatScriptField INSTANCE = new ToHalfFloatScriptField();
+
+        private ToHalfFloatScriptField() {
+
+        }
+
+        public DocValuesField<?> getScriptField(SortedNumericDoubleValues sortedNumericDoubleValues, String name) {
+            return new DelegateDocValuesField(new Doubles(sortedNumericDoubleValues), name);
+        }
+    }
+
+    public static class ToFloatScriptField extends ToScriptField {
+
+        public static final ToFloatScriptField INSTANCE = new ToFloatScriptField();
+
+        private ToFloatScriptField() {
+
+        }
+
+        public DocValuesField<?> getScriptField(SortedNumericDoubleValues sortedNumericDoubleValues, String name) {
+            return new DelegateDocValuesField(new Doubles(sortedNumericDoubleValues), name);
+        }
+    }
+
+    public static class ToScaledFloatScriptField extends ToScriptField {
+
+        public static final ToScaledFloatScriptField INSTANCE = new ToScaledFloatScriptField();
+
+        private ToScaledFloatScriptField() {
+
+        }
+
+        public DocValuesField<?> getScriptField(SortedNumericDoubleValues sortedNumericDoubleValues, String name) {
+            return new DelegateDocValuesField(new Doubles(sortedNumericDoubleValues), name);
+        }
+    }
+
+    public static class ToDoubleScriptField extends ToScriptField {
+
+        public static final ToDoubleScriptField INSTANCE = new ToDoubleScriptField();
+
+        private ToDoubleScriptField() {
+
+        }
+
+        public DocValuesField<?> getScriptField(SortedNumericDoubleValues sortedNumericDoubleValues, String name) {
+            return new DelegateDocValuesField(new Doubles(sortedNumericDoubleValues), name);
+        }
+    }
+
+    public static class ToDateMillisScriptField extends ToScriptField {
+
+        public static final ToDateMillisScriptField INSTANCE = new ToDateMillisScriptField();
+
+        private ToDateMillisScriptField() {
+
+        }
+
+        public DocValuesField<?> getScriptField(SortedNumericDocValues sortedNumericDocValues, String name) {
+            return new DelegateDocValuesField(new Dates(sortedNumericDocValues, false), name);
+        }
+    }
+
+    public static class ToDateNanosScriptField extends ToScriptField {
+
+        public static final ToDateNanosScriptField INSTANCE = new ToDateNanosScriptField();
+
+        private ToDateNanosScriptField() {
+
+        }
+
+        public DocValuesField<?> getScriptField(SortedNumericDocValues sortedNumericDocValues, String name) {
+            return new DelegateDocValuesField(new Dates(sortedNumericDocValues, true), name);
+        }
+    }
+
```

Comment:
I wonder if we could use generics to collapse this down to a single method? Then this could be a functional interface which would make a lot of this clearer; something like 
```
public interface ToScriptField<T> {
    public DocValuesField<?> getScriptField(T docValues, String name);

    final ToScriptField<SortedNumericDocValues> DATE_NANOS = (v, n) -> return new DelegateDocValuesField(new Dates(v, true), n);
}
```

Reply:
See top level comment.